### PR TITLE
OW-205 - add on() callback to media processor

### PR DESCRIPTION
**Pull request by gmininberg**

### CODE REVIEW

File: lib/src/core/MediaProcessor.ts [hunk 1]
```diff
@@ -1,7 +1,9 @@
-import Pipeline from './pipeline'
+import Pipeline, { EventDataMap } from './pipeline'
 import { Key } from '../telemetry/Key'
 import { Report, ReportBuilder, Reporter } from '../telemetry/Reporter';
 import { v4 as uuid } from 'uuid';
+import Emittery from 'emittery'
+
 
 /**
  * Media processor class holding and running the media processing logic.
```
Comment (joliveraortega) [inline lib/src/core/MediaProcessor.ts:9]: We should mention that this class implements an async event emitter pattern now.

Comment (joliveraortega) [inline lib/src/core/MediaProcessor.ts:9]: We should mention that this class implements an async event emitter pattern now.

File: lib/src/core/MediaProcessor.ts [hunk 2]
```diff
@@ -13,9 +15,11 @@ import { v4 as uuid } from 'uuid';
  *   let transformers: Array<Transformer> = [];
  *   transformers.push(new CanvasTransform());
  *   mediaProcessor.setTransformers(transformers);
+ *   mediaProcessor.on('error',(ErrorData => {}))
+ *   mediaProcessor.on('error',(WarnData => {}))
```
Comment (joliveraortega) [inline lib/src/core/MediaProcessor.ts:19]: Nit: I guess you did not want to say `error` there.

Comment (joliveraortega) [inline lib/src/core/MediaProcessor.ts:19]: Nit: I guess you did not want to say `error` there.

File: lib/src/core/MediaProcessor.ts [hunk 3]
```diff
@@ -47,6 +59,23 @@ class MediaProcessor {
...
 
+  /**
+   * sets the expected rate of the track per second
+   * the media processor will use this number for calculating drops in the rate.
+   * this could happen when transform will take more time then expected.
+   * this will not cause an error just warning to the client.
+   * mostly:
+   * video: 30 frames per second
+   * audio: 50 audio data per second for OPUS
+   * @param trackExpectedRate
```
Comment (joliveraortega) [inline lib/src/core/MediaProcessor.ts:70]: Nit: Starts sentences with capital letters and format the whole comment in a single paragraph please. Document the param also please.

Comment (joliveraortega) [inline lib/src/core/MediaProcessor.ts:70]: Nit: Starts sentences with capital letters and format the whole comment in a single paragraph please. Document the param also please.

File: lib/src/core/pipeline.ts [hunk 4]
```diff
@@ -1,11 +1,79 @@
...
+ */
+export type DropInfo = {
+  requested: number
+  current: number
+}
+
+/**
+ * ErrorPostion specifies the function which the error (exception) happened
+ */
+export type ErrorPostion = 'start' | 'transform' | 'flush'
```
Comment (joliveraortega) [inline lib/src/core/pipeline.ts:24]: `ErrorFunction` maybe. I know you love wording change suggestions.

Comment (joliveraortega) [inline lib/src/core/pipeline.ts:24]: `ErrorFunction` maybe. I know you love wording change suggestions.

File: lib/src/core/pipeline.ts [hunk 5]
```diff
@@ -1,11 +1,79 @@
...
+
+/**
+ * EventMetaData the meta data of the event.
+ * transformerIndex is the transformer index in the array of transformers.
```
Comment (joliveraortega) [inline lib/src/core/pipeline.ts:28]: Member are documented in the member itself (not here).

Comment (joliveraortega) [inline lib/src/core/pipeline.ts:28]: Member are documented in the member itself (not here).

File: lib/src/core/pipeline.ts [hunk 6]
```diff
@@ -1,11 +1,79 @@
...
+export type EventMetaData = {
+  transformerIndex: number
+};
+
+/**
+ * WarnData the warning data type
+ * ```ts
+ *  {
+ *    eventMetaData: { transformerIndex: 0},
+ *    warningType: WarningType.FPS_DROP
```
Comment (joliveraortega) [inline lib/src/core/pipeline.ts:39]: nit: `dropInfo` is missing here.

Comment (joliveraortega) [inline lib/src/core/pipeline.ts:39]: nit: `dropInfo` is missing here.

File: lib/src/core/pipeline.ts [hunk 7]
```diff
@@ -1,11 +1,79 @@
...
+ *    eventMetaData: { transformerIndex: 0},
+ *    ErrorData: 'start',
+ *    error: e (the exception in the catch)
+ *  }
+ * ```
+ */
+export type ErrorData = {
+  eventMetaData: EventMetaData
+  position: ErrorPostion
+  error: unknown
```
Comment (joliveraortega) [inline lib/src/core/pipeline.ts:62]: What's the type for `error`?, Is it `unknown` a type?

Comment (joliveraortega) [inline lib/src/core/pipeline.ts:62]: What's the type for `error`?, Is it `unknown` a type?

File: lib/src/core/pipeline.ts [hunk 8]
```diff
@@ -85,6 +164,8 @@ class InternalTransformer implements Transformer {
             .variation('Error')
             .build();
           Reporter.report(report);
+          let msg: ErrorData = {eventMetaData: {transformerIndex: this.index_}, error: e, position: 'transform'}
```
Comment (joliveraortega) [inline lib/src/core/pipeline.ts:167]: Ditto

File: lib/src/core/pipeline.ts [hunk 9]
```diff
@@ -108,6 +189,8 @@ class InternalTransformer implements Transformer {
           .variation('Error')
           .build();
         Reporter.report(error);
+        let msg: ErrorData = {eventMetaData: {transformerIndex: this.index_}, error: e, position: 'flush'}
```
Comment (joliveraortega) [inline lib/src/core/pipeline.ts:192]: Ditto

Comment (joliveraortega) [inline lib/src/core/pipeline.ts:192]: Ditto

File: lib/src/core/pipeline.ts [hunk 10]
```diff
@@ -129,6 +212,11 @@ class InternalTransformer implements Transformer {
     let timeElapsed_s = ((Date.now() - this.mediaTransformerQosReportStartTimestamp_) / 1000);
     let fps: number = this.framesFromSource_ / timeElapsed_s;
     let transformedFps: number = this.framesTransformed_ / timeElapsed_s;
+    console.log(fps, transformedFps, this.trackExpectedRate_);
```
Comment (joliveraortega) [inline lib/src/core/pipeline.ts:215]: nit: Do we need this?

Comment (joliveraortega) [inline lib/src/core/pipeline.ts:215]: nit: Do we need this?

File: lib/src/core/pipeline.ts [hunk 11]
```diff
@@ -129,6 +212,11 @@ class InternalTransformer implements Transformer {
...
+    if(this.trackExpectedRate_ != -1 && this.trackExpectedRate_ * 0.8 > fps){
```
Comment (joliveraortega) [inline lib/src/core/pipeline.ts:216]: What's `0.8`? Maybe It should be moved to a const var with some meaningful name.

Comment (joliveraortega) [inline lib/src/core/pipeline.ts:216]: What's `0.8`? Maybe It should be moved to a const var with some meaningful name.

File: lib/src/core/pipeline.ts [hunk 12]
```diff
@@ -129,6 +212,11 @@ class InternalTransformer implements Transformer {
...
+      let msg: WarnData = {eventMetaData: {transformerIndex: this.index_}, warningType: WarningType.FPS_DROP, dropInfo: {requested: this.trackExpectedRate_, current: fps}}
```
Comment (joliveraortega) [inline lib/src/core/pipeline.ts:217]: Ditto

Comment (joliveraortega) [inline lib/src/core/pipeline.ts:217]: Ditto

**Review by joliveraortega — COMMENTED** [review]
Looking good. Almost there, left some comments/nits.

File: lib/src/core/pipeline.ts (hunk 7 above)
Comment (gmininberg) [reply lib/src/core/pipeline.ts:62]: yes... thats the type we of catch. it can be anything

**Pull request merged**